tests/s3_test.go

```diff
@@ -0,0 +1,85 @@
+package tests
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/gruntwork-io/terratest/modules/aws"
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/gruntwork-io/terratest/modules/random"
+	"github.com/gruntwork-io/terratest/modules/retry"
+	"github.com/gruntwork-io/terratest/modules/terraform"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTerraformS3Bucket(t *testing.T) {
+	t.Parallel()
+
+	expectedName := fmt.Sprintf("terratest-aws-s3-example-%s", strings.ToLower(random.UniqueId()))
+
+	awsRegion := aws.GetRandomStableRegion(t, nil, nil)
+
+	terraformOptions := terraform.WithDefaultRetryableErrors(t, &terraform.Options{
+		TerraformDir: "../examples/test_minimal",
+
+		Vars: map[string]interface{}{
+			"test_bucket_name": expectedName,
+			"aws_region":       awsRegion,
+		},
+
+		EnvVars: map[string]string{
+			"AWS_DEFAULT_REGION": awsRegion,
+		},
+	})
+
+	defer terraform.Destroy(t, terraformOptions)
+
+	terraform.InitAndApply(t, terraformOptions)
+
+	b, err := terraform.OutputMapOfObjectsE(t, terraformOptions, "test-bucket")
+
+	if err != nil {
+		logger.Log(t, err)
+	}
+
+	bucket_name := b["bucket_name"].(string)
+
+	assert.Equal(t, expectedName, bucket_name)
+	aws.AssertS3BucketPolicyExists(t, awsRegion, bucket_name)
+
+	// Test Unencrypted File
+	bodyString := "test-body"
+	k := fmt.Sprintf("example-file-%s", strings.ToLower(random.UniqueId()))
+	upParams := &s3manager.UploadInput{
+		Bucket: &bucket_name,
+		Key:    &k,
+		Body:   strings.NewReader(bodyString),
+	}
+
+	up := aws.NewS3Uploader(t, awsRegion)
+
+	_, err = up.Upload(upParams)
+	assert.Error(t, err)
+
+	// Test AES256 Encrypted File
```

Comment:
I'm obviously not the expert here, but it seems like this test covers a number of cases.

What's the lifecycle like for this test framework?

I'm wondering if we can break up these cases into their own functions.

Ideally would like the flow to be something like:
1. Create test infrastructure
2. Test 1
3. Test 2
4. ...
5. Test n
6. Teardown testinfrastructure

That way when we're looking at the test results, we'll know what cases pass and fail.

Reply:
We definitely can! There are a few ways we can do that. I'll be adding in next commits and I ask for your review again. Thanks.